Doc/c-api/bytes.rst

```diff
@@ -201,3 +201,52 @@ called with a non-bytes parameter.
    reallocation fails, the original bytes object at *\*bytes* is deallocated,
    *\*bytes* is set to ``NULL``, :exc:`MemoryError` is set, and ``-1`` is
    returned.
+
+PyBytesWriter
+^^^^^^^^^^^^^
+
+The :c:type:`PyBytesWriter` API can be used to create a Python :class:`bytes`
+object.
+
+.. versionadded:: 3.14
+
+.. c:type:: PyBytesWriter
+
+   A bytes writer instance.
+
+   The instance must be destroyed by :c:func:`PyBytesWriter_Finish` on
+   success, or :c:func:`PyBytesWriter_Discard` on error.
+
+.. c:function:: PyBytesWriter* PyBytesWriter_Create(Py_ssize_t size, char **str)
+
+   Create a bytes writer instance.
+   Preallocate *size* bytes.
+
+   On success, set *\*str* and return a new writer.
+   On error, set an exception and return ``NULL``.
+
+.. c:function:: PyObject* PyBytesWriter_Finish(PyBytesWriter *writer, char *str)
+
+   Return the final Python :class:`bytes` object and destroy the writer
+   instance.
+
+   On success, return a bytes object.
+   On error, set an exception and return ``NULL``.
+
+.. c:function:: void PyBytesWriter_Discard(PyBytesWriter *writer)
+
+   Discard the internal bytes buffer and destroy the writer instance.
+
+.. c:function:: int PyBytesWriter_Prepare(PyBytesWriter *writer, char **str, Py_ssize_t size)
+
+   Allocate *size* bytes to prepare writing *size* bytes into *writer*.
+
+   On success, update *\*str* and return ``0``.
+   On error, set an exception and return ``-1``.
+
+.. c:function:: int PyBytesWriter_WriteBytes(PyBytesWriter *writer, char **str, const void *bytes, Py_ssize_t size)
```

Comment:
Do we need to always carry over \*\*str, or is it possible to have the *writer* instance holds a reference to it and be responsible for it (like for PyUnicode_Writer?) Or is the current interface not compatible with this kind of approach?

Reply:
When I designed the private _PyBytesWriter API, I tried different approaches. This one ("carry str") is the most efficient: https://vstinner.github.io/pybyteswriter.html

Reply:
Would it make sense to check if after (almost) 10 years this has been solved? I don't know when you checked the last time whether the machine code being produced is unoptimized, but are you interested in me trying to figure out something? Or do you want to keep the private API as it is?

Reply:
So I checked for GCC 14 and Clang 18:

```C
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char * restrict str;
    char  buffer[100];
} writer_t;

void init(writer_t *w) { w->str = w->buffer; }

static char* __attribute__((used)) 
f(int n)
{
    writer_t writer;
    init(&writer);
    while (1) {
        *(writer.str)++ = 'a';
    }
    *writer.str = '\0';
    return strdup(writer.buffer);
}

static char * __attribute__((used))
g(int n)
{
    writer_t writer;
    init(&writer);
    char *str = writer.str;
    while (1) {
        *str++ = 'a';
    }    
    *str = '\0';
    return strdup(writer.buffer);
}
```

I used `while(1)` just to force the compiler not to optimize using memset. The current version is `g` and the ideal one would be `f`. Here is the ASM output of gcc:

```asm
f:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
.L4:    ; while (1) in f()
        mov     rax, QWORD PTR [rsp]  
        lea     rdx, [rax+1]		  
        mov     QWORD PTR [rsp], rdx  
        mov     BYTE PTR [rax], 97    
        jmp     .L4
g:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
        mov     rax, QWORD PTR [rsp]
.L7:    ; while (1) in g()
        add     rax, 1
        mov     BYTE PTR [rax-1], 97
        jmp     .L7
```

So, yes, we still have the useless

```asm
mov     rax, QWORD PTR [rsp]  
lea     rdx, [rax+1]		  
mov     QWORD PTR [rsp], rdx  
mov     BYTE PTR [rax], 97
```

which would have been equivalent to

```asm
        mov     rax, QWORD PTR [rsp]
.L7:    ; while (1) in g()
        add     rax, 1
        mov     BYTE PTR [rax-1], 97
        jmp     .L7
```

And what's funny (it's not) is that using gcc 4.4.7 actually produces the following output:

```asm
init:
        lea     rax, [rdi+8]
        mov     QWORD PTR [rdi], rax
        ret
f:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
        mov     rax, QWORD PTR [rsp]
.L4:
        mov     BYTE PTR [rax], 97
        add     rax, 1
        jmp     .L4
g:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
        mov     rax, QWORD PTR [rsp]
.L8:
        mov     BYTE PTR [rax], 97
        add     rax, 1
        jmp     .L8
```

So I think we even regressed... Well, I agree that here it's an infinite loop so it's not really important but at least in v4, it was smart to put 

```asm
mov     rax, QWORD PTR [rsp]
``` 

outside of the loop... In the recent versions of GCC, it's not even the case =/

Reply:
> So, yes, we still have the useless: (...)

I'm not surprised. You can compare by moving `char *str` outside the structure to check if the machine code looks more efficient.